File: problems/988_smallest_string_starting_from_leaf.py
from typing import Self


class TreeNode:
    def __init__(
        self, val: int = 0, left: Self | None = None, right: Self | None = None
    ):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def smallestFromLeaf(self, root: TreeNode | None) -> str:
        if root is None:
            return ""

        smallest: list[int] = []

        def traverse(node: TreeNode | None, acc: list[int] = []) -> None:
            nonlocal smallest
            if node is None:
                return

            is_leaf = node.left is None and node.right is None
            digits = [node.val, *acc]

            traverse(node.left, digits)
            traverse(node.right, digits)

            if is_leaf:
                if not smallest:
                    smallest = digits
                else:
                    min_length = min(len(smallest), len(digits))
                    for i in range(min_length):
                        if digits[i] < smallest[i]:
                            smallest = digits
                            return
                        if digits[i] > smallest[i]:
                            return

                    # If we reach here, it means all the characters were the same so
                    # take the shorter of the 2 strings.
                    if len(digits) < len(smallest):
                        smallest = digits

        traverse(root)
        chars = [chr(ord("a") + digit) for digit in smallest]
        return "".join(chars)

File: problems/988-smallest-string-starting-from-leaf.ts
export {}

class TreeNode {
  val: number
  left: TreeNode | null
  right: TreeNode | null

  constructor(
    val: number = 0,
    left: TreeNode | null = null,
    right: TreeNode | null = null
  ) {
    this.val = val
    this.left = left
    this.right = right
  }
}

const CHAR_CODES_MAP: { [index: number]: string } = "abcdefghijklmnopqrstuvwxyz"
  .split("")
  .reduce((acc, char, i) => ({ ...acc, [i]: char }), {})

function smallestFromLeaf(root: TreeNode | null): string {
  let smallest: number[] = []

  function traverse(node: TreeNode | null, acc: number[]): void {
    if (node) {
      const digits = [node.val, ...acc]
      const isLeaf = node.left === null && node.right === null

      traverse(node.left, digits)
      traverse(node.right, digits)

      if (isLeaf) {
        if (smallest.length === 0) {
          smallest = digits
        } else {
          const minLength = Math.min(smallest.length, digits.length)
          for (let i = 0; i < minLength; i++) {
            if (digits[i] < smallest[i]) {
              smallest = digits
              return
            } else if (digits[i] > smallest[i]) {
              return
            }
          }

          // If we got here, it means one of the array is a prefix of the other, so set
          // smallest to be the shorter of the 2.
          if (digits.length < smallest.length) {
            smallest = digits
          }
        }
      }
    }
  }

  traverse(root, [])
  const str = smallest.map((code) => CHAR_CODES_MAP[code]).join("")
  return str
}

const n0 = new TreeNode(4)
const n1 = new TreeNode(0)
const n2 = new TreeNode(1)
const n3 = new TreeNode(1)

n0.left = n1
n0.right = n2
n1.left = n3

console.log(smallestFromLeaf(n0))
